CMS API: Pages: Fix update section

jlledom opened the pull request:
**What this PR does / why we need it**:

There's a bug in the CMS templates controller that prevents updating a CMS section for a page using the API.

Besides, I refactored the code a bit to handle the sections and the layouts using the same design, it was a bit confusing for me when I saw the old code the first time.

I also added some tests to ensure all combinations are working fine.

**Which issue(s) this PR fixes** 

This is part of [THREESCALE-8991](https://issues.redhat.com/browse/THREESCALE-8991)

**Verification steps** 

The next request should update the section for the given page:

```
curl --request PUT \
  --url 'http://provider-admin.3scale.localhost:3000/admin/api/cms/templates/$ID?access_token=$TOKEN' \
  --header 'Content-Type: application/json' \
  --data '{
        "type": "page",
        "title": "test API2",
        "path": "/testUI2",
        "section_name": "application aºlerts"
}                                           
'

```  

**Special notes for your reviewer**:

These are the possible parameter combinations and the value they should write in the DB:

**Section:**
|| #create  | #update |
|------| ------------- | ------------- |
|**id**: nil, **name**: nil| root  | previous_value  |
|**id**: "", **name**: nil| root  | previous_value  |
|**id**: nil, **name**: ""| root  | previous_value  |
|**id**: nil, **name**: name_value| name_value  | name_value  |
|**id**: id_value, **name**: nil| id_value  | id_value  |
|**id**: id_value, **name**: name_value| id_value  | id_value  |



**Layout:**
|| #create  | #update |
|------| ------------- | ------------- |
|**id**: nil, **name**: nil| nil  | previous_value  |
|**id**: "", **name**: nil| nil  | nil  |
|**id**: nil, **name**: ""| nil  | nil  |
|**id**: nil, **name**: name_value| name_value  | name_value  |
|**id**: id_value, **name**: nil| id_value  | id_value  |
|**id**: id_value, **name**: name_value| id_value  | id_value  |

mayorova commented on app/controllers/admin/api/cms/templates_controller.rb:
```diff
@@ -170,12 +168,9 @@ def find_section
-    end
+  def find_layout
+    scope = current_account.layouts
+    scope.find_by_id(params[:layout_id]) || scope.find_by_system_name(params[:layout_name])
```
We're querying the DB even when `layout_id` and `layout_name` are nil (in fact, if none of them is provided, we're still adding 2 extra queries), which is not nice, I think.

mayorova replied:
But the original implementation is indeed confusing, so I am up for refactoring.

mayorova replied:
I now also noticed the same logic already existed for sections (in `find_section`), I think we can change it also.

jlledom replied:
Done!

mayorova commented on app/controllers/admin/api/cms/templates_controller.rb:
```diff
@@ -104,7 +102,12 @@ def show
+      section = find_section
+      @template.section = section if section.present?
+    end
+    if @template.respond_to?(:layout)
```
CodeClimate complains about all `respond_to?` statements.

Maybe indeed we can refactor it to either:
- delegate the params update to the corresponding subclasses
- update the params in the controller, but before filter out the ones that shouldn't (or couldn't) be updated in the current instance (due to its class)

jlledom replied:
Done. I refactored to remove all business logic from the controller and move it to its own services. I might have over-refactored 🫠

mayorova commented on app/services/admin/api/cms/template_service.rb:
```diff
@@ -0,0 +1,99 @@
+      return unless type == CMS::Page
+
+      section = find_section
+      template.section = section if section.present?
```
That piece of logic is not right.
As the `find_section` always returns something (i.e. the Root section, in case section-related params are missing), an existing section will be overriden to Root in case neither `section_id` or `section_name` are provided, and we don't want that.

jlledom replied:
You're right, I fixed it and also fixed the test, which was wrong.

mayorova commented on app/controllers/admin/api/cms/templates_controller.rb:
```diff
@@ -2,8 +2,7 @@ class Admin::Api::CMS::TemplatesController < Admin::Api::CMS::BaseController
 
-  ALLOWED_PARAMS = %i[type system_name title path draft section_id layout_name
-                      layout_id liquid_enabled handler content_type].freeze
+  ALLOWED_PARAMS = %i[type system_name title path draft liquid_enabled handler content_type section_id section_name layout_id layout_name].freeze
```
Hard to see in the diff, but the change is that `:section_name` was removed.

mayorova commented on app/services/admin/api/cms/template_service.rb:
```diff
@@ -0,0 +1,97 @@
+    private
+
+    def attach_section_layout
+      return unless type == CMS::Page
```
It's a bit confusing, that both Update and Create classes have a `:type` attribute, but in one case it's a class, but in the other it's a symbol. Maybe I'd prefer to choose different names so there is no confusion...

jlledom replied:
That's because the `#update` request doesn't accept `type` as parameter, so I have to guess the type from the class. I renamed the attribute.

mayorova commented on app/services/admin/api/cms/template_service.rb:
```diff
@@ -0,0 +1,97 @@
+  end
+
+  class Update < Admin::Api::CMS::TemplateService
+    def initialize(template, current_account, params)
```
I would put template as the 3rd argument, so the first two are in the same order in the parent and "sibling" classes. But it's just my personal opinion, maybe not a good one.

jlledom replied:
Perfect, I pushed this

mayorova reviewed (changes requested):
This might not be this PRs responsibility, but as this part is being refactored, I think it's relevant.

There is a use case that is currently NOT supported by the API, which is - setting layout to nil (no layout defined) on **update**.

We're leaving the current layout if we can't find the layout. But I think in this case we need to distinguish two cases:
1. `layout_id` and `layout_name` are not present at all
2. some is present, but is empty (e.g. `layout_id=&title=whatever`)
In case 2. I guess we'd need to remove the layout association.


Another thing is that both for sections and layouts if we can't find the section or layout with the provided ID/name, we just ignore them.

I am thinking that this behavior might not be right - the user will get a 200 OK and thing that all was updated correctly, while it's not the case (and you won't know until you check the returned object). Maybe we should throw an error (404 Not Found, I guess), when the section / layout ID or name are provided, but there is no section / layout found with this ID / name.

jlledom commented:
@mayorova I added the two features you requested. I think now the four possibilities are covered:

- param is not sent
- param is sent empty
- params is sent with a non-existing value
- param is sent with a valid value

For both layouts and sections. I wrote tests for all. I also added a new test suite for the service, which I think is the correct place to test this logic. Since there were already tests for the controller that checked this layouts/sections behavior, I updated them too, but I think we should remove them. The tests for the controller are testing the same as the tests for the service, and the later are in the in the place where the logic is actually happening.

mayorova commented on app/services/admin/api/cms/template_service.rb:
```diff
@@ -73,10 +73,19 @@ def attach_section_layout
+      if layout.present?
+        template.layout = layout
+      elsif layout_received_empty?
+        # We received the parameter, but empty. So the user explicitly want's to remove the layout
```
`wants` :grimacing:

jlledom replied:
fixed! 🤦🏼‍♂️
